tests/integration/sandbox.py

```diff
@@ -718,6 +722,227 @@ def restart(self):
+        elif self.op_type == "read":
+            return f"[:r \"{self.key}\" nil]"
+        else:
+            raise Exception(f"invalid op_type = {self.op_type}")
+
+
+class Elle(object):
+    def __init__(self, config):
+        self.config = config
+        workdir = os.path.abspath(self.config.workdir)
+        self.logdir = os.path.join(workdir, str(uuid.uuid4()))
+        os.makedirs(self.logdir, exist_ok=True)
+        self.lock = threading.Lock()
+        self.logfile = open(os.path.join(self.logdir, "logfile.edn"), "x")
+        self.index = {}
+        self.value = 0
+
+    @staticmethod
+    def generate_ops_value(ops: typing.List[ElleOp]) -> str:
+        val = " ".join(map(str, ops))
+        return f"[{val}]"
+
+    def generate_command_output(self, thread: int, ops: List[ElleOp], state: str) -> str:
+        return f"{{:index {self.index[thread]} :process {thread} :type :{state}, :value {self.generate_ops_value(ops)}}}"
+
+    def log_command(self, thread: int, ops: List[ElleOp]):
+        try:
+            self.index[thread] += 1
+        except KeyError:
+            self.index[thread] = 1
+
+        output = self.generate_command_output(thread, ops, "invoke")
+        self.lock.acquire()
+        self.logfile.write(output)
+        self.logfile.write("\n")
+        self.lock.release()
+
+    @staticmethod
+    def generate_results_value(results: typing.List, ops: typing.List[ElleOp]) -> str:
+        val = "["
+        for i in range(len(ops)):
+            if i != 0:
+                val += " "
+            if ops[i].op_type == "append":
+                val += str(ops[i])
+            elif ops[i].op_type == "read":
+                tmp = " ".join([x.decode('utf-8') for x in results[i]])
+                val += f"[:r \"{ops[i].key}\" [{tmp}]]"
+            else:
+                raise Exception(f"unknown op type {ops[i].op_type}")
+        val += "]"
+        return val
+
+    def generate_result_output(self, thread: int, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        output: str
+        if result == "ok":
+            output = self.generate_results_value(results, ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        elif result == "fail":
+            output = self.generate_ops_value(ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        else:
+            raise Exception(f"unknown result value {result}")
+
+    def log_result(self, thread, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        # if this throws an exception, we have a failure elsewhere
+        self.index[thread] += 1
+
+        output = self.generate_result_output(thread, result, results, ops)
+        self.lock.acquire()
+        self.logfile.write(output)
+        self.logfile.write("\n")
+        self.lock.release()
+
+    def log_comment(self, msg: str):
```

Comment:
Is this unused?

Reply:
for now, the thought was that it could add comment lines to the log file produced for elle so could add "marks" so can understand output better.